### ISSUE TITLE
SSL_new(ctx) does not copy hostflags from SSL_CTX to SSL

### ISSUE THREAD

**Issue by tiran** [openssl/openssl#14579]
Python's ssl module uses ``ssl.SSLContext`` objects are central configuration objects for all TLS related options. The C code behind the ``SSLContext`` object is mostly a wrapper of ``SSL_CTX``.

A while ago I added a setter/getter for ``X509_CHECK_FLAG_NEVER_CHECK_SUBJECT`` flag. 

```python
    @property
    def hostname_checks_common_name(self):
        ncs = self._host_flags & _ssl.HOSTFLAG_NEVER_CHECK_SUBJECT
        return ncs != _ssl.HOSTFLAG_NEVER_CHECK_SUBJECT

    @hostname_checks_common_name.setter
    def hostname_checks_common_name(self, value):
        if value:
            self._host_flags &= ~_ssl.HOSTFLAG_NEVER_CHECK_SUBJECT
        else:
            self._host_flags |= _ssl.HOSTFLAG_NEVER_CHECK_SUBJECT```
```

```C
static int
set_host_flags(PySSLContext *self, PyObject *arg, void *c)
{
    X509_VERIFY_PARAM *param;
    unsigned int new_flags = 0;

    if (!PyArg_Parse(arg, "I", &new_flags))
        return -1;

    param = SSL_CTX_get0_param(self->ctx);
    self->hostflags = new_flags;
    X509_VERIFY_PARAM_set_hostflags(param, new_flags);
    return 0;
}
```

The code correctly sets ``ctx->params->hostflags`` to value ``36`` (``X509_CHECK_FLAG_NEVER_CHECK_SUBJECT|X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS``). However it doesn't have any affect on SSLSocket ``ssl->params->hostflags`` is always ``0`` and TLS handshake accepts certificate with just a subject CN and no SAN DNS entries. It took me a while to realize that hostflags are **not** copied from ``SSL_CTX`` to ``SSL``. ``X509_VERIFY_PARAM_inherit()`` ignores the hostflags when there is no verify hosts configured. Python's ssl module uses ``X509_VERIFY_PARAM_set1_host()`` only on connection objects, never on the context itself. ``ctx->params->hosts`` is always NULL.

https://github.com/openssl/openssl/blob/081a7061f3da07318c4b0f5de67b82285630bf6b/crypto/x509/x509_vpm.c#L202-L213

This behavior is surprising. I assumed that hostflags behave like any other flag. I don't see the special case documented at https://www.openssl.org/docs/man1.1.1/man3/X509_VERIFY_PARAM_set_flags.html either.

Version: OpenSSL 1.1.1j
Python upstream bug: https://bugs.python.org/issue43522
workaround PR: https://github.com/python/cpython/pull/24899

**Comment by tiran** [comment]
I guess you cannot change the behavior in OpenSSL 1.1.1. It would be backwards incompatible. Could you please update the documentation for ``X509_VERIFY_PARAM_set_flags`` and reconsider the behavior for OpenSSL 3.0.0?

**Comment by t8m** [comment]
Hmm I'm wondering whether this should not be just treated as a bug even on 1.1.1. Or in other words, the current behavior seems to be quite surprising to me.
@vdukhovni is the original author of the code in 8abffa4a73fcbf6536e0a42d736ed9211a8204ea so maybe he could give some insight?

**Comment by tiran** [comment]
@t8m @mattcaswell @vdukhovni Are you going to treat the issue as a 1.1.1 documentation bug or an implementation bug? The issue has security implications. For example Python's ssl module sets ``X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS`` by default to disable match of certificates with partial wildcards like ``spam*.python.org``. Since the flag is not propagated from SSL_CTX* to SSL*, the setting has no affect.

**Comment by vdukhovni** [comment]
This feels like an implementation bug to me.  I always set both the flags and the host names on the SSL connection, so never noticed the issue with SSL_CTX not doing the right thing.  The context's flags should probably be cloned when the SSL handle is created.

**Comment by tiran** [comment]
@mattcaswell @t8m I'm not familiar with your backport policy. The ticket is labeled for 1.1.1 but I don't see the fix in 1.1.1 branch. Are you going to backport the fix to 1.1.1 before the next release?

**Comment by mattcaswell** [comment]
Looks like the fix was only labelled for master not 1.1.1. If its a bug fix then there should be no reason why it can't be backported. When raising a PR we can keep it all in one PR if it applies cleanly to both branches. Otherwise we'd normally do 2 PRs one for each branch. Since the master PR wasn't flagged for 1.1.1, and wasn't reviewed with that in mind, I'd suggest opening a new backport PR that specifically targets 1.1.1.
